nifi-nar-bundles/nifi-framework-bundle/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/repository/VolatileProvenanceRepository.java

```diff
@@ -0,0 +1,409 @@
+import org.apache.nifi.provenance.StandardProvenanceEventRecord;
+import org.apache.nifi.provenance.lineage.ComputeLineageSubmission;
+import org.apache.nifi.provenance.search.Query;
+import org.apache.nifi.provenance.search.QuerySubmission;
+import org.apache.nifi.provenance.search.SearchableField;
+import org.apache.nifi.util.RingBuffer;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class VolatileProvenanceRepository implements ProvenanceRepository {
+
+    // default property values
+    public static final int DEFAULT_BUFFER_SIZE = 10000;
+
+    public static String CONTAINER_NAME = "in-memory";
+
+    private final RingBuffer<ProvenanceEventRecord> ringBuffer;
+    private final int maxSize;
+
+    private final AtomicLong idGenerator = new AtomicLong(0L);
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+
+    /**
+     * Default no args constructor for service loading only
+     */
+    public VolatileProvenanceRepository() {
+        ringBuffer = null;
+        maxSize = DEFAULT_BUFFER_SIZE;
+    }
+
+    public VolatileProvenanceRepository(final int maxEvents) {
+        maxSize = maxEvents;
+        ringBuffer = new RingBuffer<>(maxSize);
+    }
+
+    @Override
+    public void initialize(final EventReporter eventReporter, final Authorizer authorizer, final ProvenanceAuthorizableFactory resourceFactory,
+                           final IdentifierLookup idLookup) throws IOException {
+        if (initialized.getAndSet(true)) {
+            return;
+        }
+    }
+
+    @Override
+    public ProvenanceEventRepository getProvenanceEventRepository() {
+        return this;
+    }
+
+    @Override
+    public ProvenanceEventBuilder eventBuilder() {
+        return new StandardProvenanceEventRecord.Builder();
+    }
+
+    @Override
+    public void registerEvent(final ProvenanceEventRecord event) {
+        final long id = idGenerator.getAndIncrement();
+        ringBuffer.add(new IdEnrichedProvEvent(event, id));
+    }
+
+    @Override
+    public void registerEvents(final Iterable<ProvenanceEventRecord> events) {
+        for (final ProvenanceEventRecord event : events) {
+            registerEvent(event);
+        }
+    }
+
+    @Override
+    public List<ProvenanceEventRecord> getEvents(final long firstRecordId, final int maxRecords) throws IOException {
+        return getEvents(firstRecordId, maxRecords, null);
+    }
+
+    @Override
+    public List<ProvenanceEventRecord> getEvents(final long firstRecordId, final int maxRecords, final NiFiUser user) throws IOException {
+        return ringBuffer.getSelectedElements(new RingBuffer.Filter<ProvenanceEventRecord>() {
+            @Override
+            public boolean select(final ProvenanceEventRecord value) {
+                if (!isAuthorized(value, user)) {
```

Comment:
Don't think we need to actually call this, given that isAuthorized will always return true

Reply:
Agree